src/content/blog/2024-08-05-rafiki-code-architecture.mdx

```diff
@@ -0,0 +1,203 @@
+
+## Introduction
+
+It has been said that the Interledger Foundation upholds open technology as one of its core values. It has also been said that [Rafiki](https://github.com/interledger/rafiki), a white-label application that enables Interledger usage, upholds this value by being open-source and allowing anyone to contribute.
+
+It has _not_ been said, however, how one comes to intimately understand Rafiki. It's not impossible, to be sure, but the Interledger endeavor has been around for long enough that a write-up at a lower level is warranted.
+
+So, what makes Rafiki tick? What do we see when we venture beyond _what_ Rafiki does and into _how_ it does those things? 
+
+### Disclaimer
+This article assumes the reader has high-level knowledge of the following concepts:
+
+* [Open Payments](https://openpayments.dev/introduction/overview/)
+* [Interledger](https://interledger.org/developers/get-started/#how-does-interledger-work)
+* [Rafiki](https://rafiki.dev/introduction/overview/)
+
+## Components
+
+Rafiki is comprised primarily of three packages:
+
+* A [`backend`](https://github.com/interledger/rafiki/tree/main/packages/backend) package that extends an API for managing Open Payments resources like Incoming or Outgoing Payments. 
+* An [`auth`](https://github.com/interledger/rafiki/tree/main/packages/auth) package that provide third parties with a method of acquiring authorization to manage Open Payments resources on the Rafiki instance's associated `backend` package.
+* A [`frontend`](https://github.com/interledger/rafiki/tree/main/packages/frontend) that serves as an Admin-level UI for that Rafiki instance. It allows the manager of that Rafiki instance to directly manage items such as other peers on the Interledger network or what currencies it supports.
+
+### The `backend` and `auth` stacks
+
+Both the `backend` and `auth` packages are largely built in the same way. They both leverage the same two Node.js frameworks:
+
+* [KoaJS](https://koajs.com/)
+* [AdonisJS](https://docs.adonisjs.com/guides/preface/introduction)
+
+KoaJS is used as the framework for setting up the API routes, assigning functions to handle business logic for those routes, and wrapping those in middlewares for addtional functionality. It's a lot like an [Express server](https://expressjs.com/) if that helps with familiarity.
+
+AdonisJS is used to perform [dependency injection](https://docs.adonisjs.com/guides/concepts/dependency-injection) on the services used by each package to perform business logic.
+
+Additionally, both packages extend a [GraphQL API](https://github.com/interledger/rafiki/pull/2828#discussion_r1704862655) that serves as an Admin API.
+
+Finally, [ObjectionJS](https://vincit.github.io/objection.js/) is used as an ORM to perform database operations.
+
+Familiarity with these frameworks will be valuable in understanding how Rafiki works, and with understanding the rest of this post.
+
+### The Rafiki Backend
```

Comment:
should we mention the ILP connector as well?

Reply:
I think we have to.